wfe/wfe.go

```diff
@@ -712,6 +731,109 @@ func (wfe *WebFrontEndImpl) UpdateAccount(
 	}
 }
 
+func (wfe *WebFrontEndImpl) verifyKeyRollover(
+	innerPayload []byte,
+	existingAcct *core.Account,
+	newKey *jose.JSONWebKey,
+	request *http.Request) *acme.ProblemDetails {
+	var innerContent struct {
+		Account string
+		OldKey  jose.JSONWebKey
+	}
+	err := json.Unmarshal([]byte(innerPayload), &innerContent)
+	if err != nil {
+		return acme.MalformedProblem("Error unmarshaling key roll-over inner JWS body")
+	}
+
+	// Check account ID
+	prefix := wfe.relativeEndpoint(request, acctPath)
+	if !strings.HasPrefix(innerContent.Account, prefix) {
+		return acme.MalformedProblem(fmt.Sprintf("Key ID (account) in inner JWS body missing expected URL prefix (provided account value: %q)", innerContent.Account))
+	}
+	accountID := strings.TrimPrefix(innerContent.Account, prefix)
+	if accountID == "" {
+		return acme.MalformedProblem(fmt.Sprintf("No key ID (account) in inner JWS body (provided account value: %q)", innerContent.Account))
+	}
+	if accountID != existingAcct.ID {
+		return acme.MalformedProblem(fmt.Sprintf("Key roll-over inner JWS body contains wrong account ID (provided account value: %q)", innerContent.Account))
+	}
+
+	// Verify inner key
+	if !keyDigestEquals(innerContent.OldKey, *existingAcct.Key) {
+		return acme.MalformedProblem("Key roll-over inner JWS body JSON contains wrong old key")
+	}
+
+	// Check for same key
+	if keyDigestEquals(innerContent.OldKey, newKey) {
+		return acme.MalformedProblem("New and old key are identical")
+	}
+
+	return nil
+}
+
+func (wfe *WebFrontEndImpl) KeyRollover(
+	ctx context.Context,
+	logEvent *requestEvent,
+	response http.ResponseWriter,
+	request *http.Request) {
+	// Extract and parse outer JWS, and retrieve account
+	body, outerHeaderURL, key, prob := wfe.verifyPOST(ctx, logEvent, request, wfe.lookupJWK)
+	if prob != nil {
+		wfe.sendError(prob, response)
+		return
+	}
+
+	existingAcct, prob := wfe.getAcctByKey(key)
+	if prob != nil {
+		wfe.sendError(prob, response)
+		return
+	}
+
+	// Extract inner JWS
+	parsedInnerJWS, err := wfe.parseJWS(string(body))
+	if err != nil {
+		wfe.sendError(acme.MalformedProblem(err.Error()), response)
+		return
+	}
+
+	newPubKey, prob := wfe.extractJWK(request, parsedInnerJWS)
+	if prob != nil {
+		wfe.sendError(prob, response)
+		return
+	}
+
+	innerPayload, innerHeaderURL, prob := wfe.verifyJWSLight(newPubKey, parsedInnerJWS, "Inner ")
+	if prob != nil {
+		wfe.sendError(prob, response)
+		return
+	}
+
+	if innerHeaderURL != outerHeaderURL {
+		wfe.sendError(acme.MalformedProblem("JWS header parameter 'url' differs for inner and outer JWS."), response)
```

Comment:
I think you missed a `return` after this `wfe.sendError` and control flow will continue when the inner header and outer header don't match

Reply:
Definitely! Thanks for noticing this!